Blockchain/task/src/blockchain/management/MoneyHandler.java

```diff
@@ -0,0 +1,120 @@
+package blockchain.management;
+
+import blockchain.exceptions.EncryptionException;
+import blockchain.exceptions.IllegalTransactionArgumentException;
+
+import java.security.*;
+import java.util.*;
+
+import static blockchain.utils.Constants.*;
+import static blockchain.utils.Encryption.getRandomNumber;
+
+public abstract class MoneyHandler extends Thread {
+    protected String nickName;
+    protected final BlockChain blockChain;
+    protected int moneyForCreation = 100;
+    private final LinkedList<Transaction> transactions = new LinkedList<>();
+
+    public MoneyHandler(String name, BlockChain blockChain) {
+        this.nickName = name;
+        this.blockChain = blockChain;
+    }
+
+    public Transaction createTransaction(String[] data) throws IllegalTransactionArgumentException {
+        MoneyHandler receiver;
+        if(data.length != 2 || ((receiver = getReceiver(data[1])) == null))
+            throw new IllegalTransactionArgumentException("Invalid input");
+        int sum = Integer.parseInt(data[0]);
+
+        checkMoney(sum);
+
+        KeyPair pair = getKeyPair();
+        Transaction transaction
+                = new Transaction(blockChain.getId(), sum, this, receiver, pair.getPublic());
+        transaction.setSignature(createSignature(transaction.toString(), pair));
+
+        this.transactions.add(transaction);
+        receiver.transactions.add(transaction);
+        blockChain.addTransaction(transaction);
+        return transaction;
+    }
+
+    public String name() {
+        return nickName;
+    }
+
+    protected KeyPair getKeyPair() {
+        try {
+            KeyPairGenerator generator = KeyPairGenerator.getInstance(DSA);
+            generator.initialize(KILOBYTE_TO_BYTE);
+            return generator.generateKeyPair();
+        } catch (NoSuchAlgorithmException e) {
+            throw new EncryptionException("Keys weren't generated");
+        }
+    }
+
+    protected int getMoneyHeld() {
+        int moneyHeld = moneyForCreation;
+
+        for (Transaction transaction : (LinkedList<Transaction>) transactions.clone()) {
+            if(transaction.getSender().equals(nickName))
+                moneyHeld -= transaction.getCoins();
+            if(transaction.getReceiver().equals(nickName))
+                moneyHeld += transaction.getCoins();
+        }
+
+        return moneyHeld;
+    }
+
+    protected byte[] createSignature(String message, KeyPair pair) {
+        try {
+            Signature sign = Signature.getInstance(SHA_WITH_DSA);
```

Comment:
same for this object and for this method - extract to separate utility class and create object only once